plutus-core/plutus-core/src/PlutusCore/Name/UniqueSet.hs

```diff
@@ -0,0 +1,91 @@
+module PlutusCore.Name.UniqueSet (
+  UniqueSet (..),
+  insertByUnique,
+  insertByName,
+  singletonName,
+  fromFoldable,
+  fromUniques,
+  fromNames,
+  memberByUnique,
+  memberByName,
+  notMemberByName,
+  (\\),
+  union,
+  setOfByUnique,
+  setOfByName,
+) where
+
+import Control.Lens (Getting, view)
+import Control.Lens.Getter (views)
+import Data.Coerce (Coercible, coerce)
+import Data.Foldable (foldl')
+import Data.IntSet qualified as IS
+import Data.IntSet.Lens qualified as IS
+import PlutusCore.Name.Unique (HasUnique (..), Unique (Unique))
+
+-- | A set containing uniques.
+newtype UniqueSet unique = UniqueSet
+  { unUniqueSet :: IS.IntSet
+  }
+  deriving newtype (Show, Eq, Semigroup, Monoid)
+
+-- | Insert a value by a unique.
+insertByUnique ::
+  (Coercible unique Unique) =>
+  unique ->
+  UniqueSet unique ->
+  UniqueSet unique
+insertByUnique = coerce . IS.insert . coerce
```

Comment:
last coerce is unnecessary?

Reply:
It's necessary because it coerces the `unique` into the `Key` type used by `IntSet`.

Reply:
Sorry, that was very amibguous! I meant the one that is applied last i.e. the one that is applied first :sweat_smile: I think we get back a `UniqueSet unique` from `insert`, so no need to coerce it again.

Reply:
Oh, that's the `IntSet` insert, its result needs to be wrapped into a `UniqueSet`.